File: experiments-local/experiment-16/runs/agent=omp_language=typescript_model=openrouter/deepseek/deepseek-v3.2_tooling=none/rep1/src/csv-loader.ts
/**
 * Brazilian Soccer MCP Server - CSV Loading Utilities
 * 
 * Loads and parses soccer datasets from CSV files with proper type handling
 * and date format normalization.
 */

import { readFileSync } from 'fs';
import { parse } from 'papaparse';
import { parse as parseDate, isValid, parseISO } from 'date-fns';
import { Match, Player, DatasetType } from './types';

/**
 * Normalizes team names by removing state suffixes and standardizing variations
 */
export function normalizeTeamName(teamName: string): string {
  if (!teamName || typeof teamName !== 'string') return '';
  
  const trimmed = teamName.trim();
  
  // Remove state suffix (e.g., "Palmeiras-SP" → "Palmeiras")
  const withoutSuffix = trimmed.replace(/-\s*[A-Z]{2}$/, '');
  
  // Handle common Brazilian club naming variations
  const variations: Record<string, string> = {
    'Sport Club Corinthians Paulista': 'Corinthians',
    'Corinthians-SP': 'Corinthians',
    'Corinthians Paulista': 'Corinthians',
    'Sociedade Esportiva Palmeiras': 'Palmeiras',
    'Palmeiras-SP': 'Palmeiras',
    'Clube de Regatas do Flamengo': 'Flamengo',
    'Flamengo-RJ': 'Flamengo',
    'São Paulo Futebol Clube': 'São Paulo',
    'São Paulo-SP': 'São Paulo',
    'Sao Paulo': 'São Paulo',
    'Grêmio Foot-Ball Porto Alegrense': 'Grêmio',
    'Gremio-RS': 'Grêmio',
    'Sport Club Internacional': 'Internacional',
    'Internacional-RS': 'Internacional',
    'Santos Futebol Clube': 'Santos',
    'Santos-SP': 'Santos',
    'Clube Atlético Mineiro': 'Atlético-MG',
    'Atletico-MG': 'Atlético-MG',
    'Cruzeiro Esporte Clube': 'Cruzeiro',
    'Cruzeiro-MG': 'Cruzeiro',
    'Fluminense Football Club': 'Fluminense',
    'Fluminense-RJ': 'Fluminense',
    'Botafogo de Futebol e Regatas': 'Botafogo',
    'Botafogo-RJ': 'Botafogo',
    'Vasco da Gama': 'Vasco',
    'Vasco da Gama-RJ': 'Vasco',
  };
  
  return variations[trimmed] || variations[withoutSuffix] || withoutSuffix || trimmed;
}

/**
 * Parses various date formats found in the datasets
 */
export function parseSoccerDate(dateStr: string): Date {
  if (!dateStr || typeof dateStr !== 'string') return new Date(NaN);
  
  const trimmed = dateStr.trim();
  
  // Try ISO format first (e.g., "2023-09-24")
  const isoDate = parseISO(trimmed);
  if (isValid(isoDate)) return isoDate;
  
  // Try Brazilian format (e.g., "29/03/2003")
  const brDate = parseDate(trimmed, 'dd/MM/yyyy', new Date());
  if (isValid(brDate)) return brDate;
  
  // Try with time component (e.g., "2012-05-19 18:30:00")
  const dateTime = parseDate(trimmed, 'yyyy-MM-dd HH:mm:ss', new Date());
  if (isValid(dateTime)) return dateTime;
  
  // Try just date part of datetime
  const datePart = trimmed.split(' ')[0];
  const dateOnly = parseISO(datePart) || parseDate(datePart, 'dd/MM/yyyy', new Date());
  if (isValid(dateOnly)) return dateOnly;
  
  // Fallback to JS Date parsing
  const fallback = new Date(trimmed);
  return isValid(fallback) ? fallback : new Date(NaN);
}

/**
 * Extracts season/year from date or explicit season field
 */
export function extractSeason(date: Date, seasonField?: string | number): number {
  if (seasonField !== undefined) {
    const num = Number(seasonField);
    if (!isNaN(num) && num > 1900 && num < 2100) return num;
  }
  
  if (isValid(date)) {
    return date.getFullYear();
  }
  
  return new Date().getFullYear();
}

/**
 * Loads and parses Brasileirão Serie A matches
 */
export function loadBrasileiraoMatches(filePath: string): Match[] {
  const content = readFileSync(filePath, 'utf-8');
  const result = parse<{
    datetime: string;
    home_team: string;
    home_team_state: string;
    away_team: string;
    away_team_state: string;
    home_goal: string;
    away_goal: string;
    season: string;
    round: string;
  }>(content, { header: true, skipEmptyLines: true });
  
  return result.data
    .filter(row => row.datetime && row.home_team && row.away_team)
    .map(row => {
      const date = parseSoccerDate(row.datetime);
      const homeGoals = Number(row.home_goal) || 0;
      const awayGoals = Number(row.away_goal) || 0;
      
      return {
        date,
        homeTeam: normalizeTeamName(row.home_team),
        awayTeam: normalizeTeamName(row.away_team),
        homeGoals,
        awayGoals,
        season: extractSeason(date, row.season),
        round: Number(row.round) || row.round,
        competition: 'Brasileirão Serie A',
        source: 'brasileirao',
        originalData: { ...row }
      };
    });
}

/**
 * Loads and parses Copa do Brasil matches
 */
export function loadCopaDoBrasilMatches(filePath: string): Match[] {
  const content = readFileSync(filePath, 'utf-8');
  const result = parse<{
    round: string;
    datetime: string;
    home_team: string;
    away_team: string;
    home_goal: string;
    away_goal: string;
    season: string;
  }>(content, { header: true, skipEmptyLines: true });
  
  return result.data
    .filter(row => row.datetime && row.home_team && row.away_team)
    .map(row => {
      const date = parseSoccerDate(row.datetime);
      const homeGoals = Number(row.home_goal) || 0;
      const awayGoals = Number(row.away_goal) || 0;
      
      return {
        date,
        homeTeam: normalizeTeamName(row.home_team),
        awayTeam: normalizeTeamName(row.away_team),
        homeGoals,
        awayGoals,
        season: extractSeason(date, row.season),
        round: row.round,
        competition: 'Copa do Brasil',
        source: 'copa-do-brasil',
        originalData: { ...row }
      };
    });
}

/**
 * Loads and parses Copa Libertadores matches
 */
export function loadLibertadoresMatches(filePath: string): Match[] {
  const content = readFileSync(filePath, 'utf-8');
  const result = parse<{
    datetime: string;
    home_team: string;
    away_team: string;
    home_goal: string;
    away_goal: string;
    season: string;
    stage: string;
  }>(content, { header: true, skipEmptyLines: true });
  
  return result.data
    .filter(row => row.datetime && row.home_team && row.away_team)
    .map(row => {
      const date = parseSoccerDate(row.datetime);
      const homeGoals = Number(row.home_goal) || 0;
      const awayGoals = Number(row.away_goal) || 0;
      
      return {
        date,
        homeTeam: normalizeTeamName(row.home_team),
        awayTeam: normalizeTeamName(row.away_team),
        homeGoals,
        awayGoals,
        season: extractSeason(date, row.season),
        stage: row.stage,
        competition: 'Copa Libertadores',
        source: 'libertadores',
        originalData: { ...row }
      };
    });
}

/**
 * Loads and parses extended match statistics
 */
export function loadExtendedStatsMatches(filePath: string): Match[] {
  const content = readFileSync(filePath, 'utf-8');
  const result = parse<{
    tournament: string;
    home: string;
    away: string;
    home_goal: string;
    away_goal: string;
    home_corner: string;
    away_corner: string;
    home_attack: string;
    away_attack: string;
    home_shots: string;
    away_shots: string;
    time: string;
    date: string;
    ht_result: string;
    at_result: string;
    total_corners: string;
  }>(content, { header: true, skipEmptyLines: true });
  
  return result.data
    .filter(row => row.date && row.home && row.away)
    .map(row => {
      // Combine date and time if available
      const dateTimeStr = row.time ? `${row.date} ${row.time}` : row.date;
      const date = parseSoccerDate(dateTimeStr);
      const homeGoals = Number(row.home_goal) || 0;
      const awayGoals = Number(row.away_goal) || 0;
      
      return {
        date,
        homeTeam: normalizeTeamName(row.home),
        awayTeam: normalizeTeamName(row.away),
        homeGoals,
        awayGoals,
        season: extractSeason(date),
        competition: row.tournament,
        homeCorners: Number(row.home_corner) || undefined,
        awayCorners: Number(row.away_corner) || undefined,
        homeAttacks: Number(row.home_attack) || undefined,
        awayAttacks: Number(row.away_attack) || undefined,
        homeShots: Number(row.home_shots) || undefined,
        awayShots: Number(row.away_shots) || undefined,
        totalCorners: Number(row.total_corners) || undefined,
        source: 'extended-stats',
        originalData: { ...row }
      };
    });
}

/**
 * Loads and parses historical Brasileirão matches (2003-2019)
 */
export function loadHistoricalMatches(filePath: string): Match[] {
  const content = readFileSync(filePath, 'utf-8');
  const result = parse<{
    Data: string;
    Ano: string;
    Rodada: string;
    Equipe_mandante: string;
    Equipe_visitante: string;
    Gols_mandante: string;
    Gols_visitante: string;
    Mandante_UF: string;
    Visitante_UF: string;
    Vencedor: string;
    Arena: string;
  }>(content, { header: true, skipEmptyLines: true });
  
  return result.data
    .filter(row => row.Data && row.Equipe_mandante && row.Equipe_visitante)
    .map(row => {
      const date = parseSoccerDate(row.Data);
      const homeGoals = Number(row.Gols_mandante) || 0;
      const awayGoals = Number(row.Gols_visitante) || 0;
      
      let winner: 'home' | 'away' | 'draw' | undefined;
      if (row.Vencedor === 'Mandante') winner = 'home';
      else if (row.Vencedor === 'Visitante') winner = 'away';
      else if (row.Vencedor === 'Empate') winner = 'draw';
      
      return {
        date,
        homeTeam: normalizeTeamName(row.Equipe_mandante),
        awayTeam: normalizeTeamName(row.Equipe_visitante),
        homeGoals,
        awayGoals,
        season: extractSeason(date, row.Ano),
        round: row.Rodada,
        stadium: row.Arena,
        winner,
        competition: 'Brasileirão Serie A',
        source: 'historical',
        originalData: { ...row }
      };
    });
}

/**
 * Loads and parses FIFA player data
 */
export function loadFifaPlayers(filePath: string): Player[] {
  const content = readFileSync(filePath, 'utf-8');
  const result = parse<{
    ID: string;
    Name: string;
    Age: string;
    Nationality: string;
    Overall: string;
    Potential: string;
    Club: string;
    Position: string;
    'Jersey Number': string;
    Height: string;
    Weight: string;
    'Preferred Foot': string;
    Value: string;
    Wage: string;
    Crossing: string;
    Finishing: string;
    HeadingAccuracy: string;
    ShortPassing: string;
    Dribbling: string;
    ShotPower: string;
    Stamina: string;
    Strength: string;
    Aggression: string;
    Composure: string;
  }>(content, { header: true, skipEmptyLines: true });
  
  const players: Player[] = [];
  
  for (const row of result.data) {
    if (!row.ID || !row.Name || !row.Nationality) continue;
    
    const id = Number(row.ID);
    if (isNaN(id)) continue;
    
    players.push({
      id,
      name: row.Name,
      age: Number(row.Age) || 0,
      nationality: row.Nationality,
      overall: Number(row.Overall) || 0,
      potential: Number(row.Potential) || 0,
      club: row.Club,
      position: row.Position,
      jerseyNumber: Number(row['Jersey Number']) || undefined,
      height: row.Height,
      weight: row.Weight,
      preferredFoot: row['Preferred Foot'],
      value: row.Value,
      wage: row.Wage,
      crossing: Number(row.Crossing) || undefined,
      finishing: Number(row.Finishing) || undefined,
      headingAccuracy: Number(row.HeadingAccuracy) || undefined,
      shortPassing: Number(row.ShortPassing) || undefined,
      dribbling: Number(row.Dribbling) || undefined,
      shotPower: Number(row.ShotPower) || undefined,
      stamina: Number(row.Stamina) || undefined,
      strength: Number(row.Strength) || undefined,
      aggression: Number(row.Aggression) || undefined,
      composure: Number(row.Composure) || undefined,
      source: 'fifa',
      originalData: { ...row }
    });
  }
  
  return players;
}

/**
 * Loads all datasets from their respective files
 */
export function loadAllDatasets(dataDir: string = 'data/kaggle'): {
  matches: Match[];
  players: Player[];
} {
  const matches: Match[] = [
    ...loadBrasileiraoMatches(`${dataDir}/Brasileirao_Matches.csv`),
    ...loadCopaDoBrasilMatches(`${dataDir}/Brazilian_Cup_Matches.csv`),
    ...loadLibertadoresMatches(`${dataDir}/Libertadores_Matches.csv`),
    ...loadExtendedStatsMatches(`${dataDir}/BR-Football-Dataset.csv`),
    ...loadHistoricalMatches(`${dataDir}/novo_campeonato_brasileiro.csv`)
  ];
  
  const players = loadFifaPlayers(`${dataDir}/fifa_data.csv`);
  
  return { matches, players };
}

File: experiments-local/experiment-opencode-brazil-kimi/runs/agent=opencode_language=csharp_model=openrouter/moonshotai/kimi-k2.7-code_tooling=none/rep2-failed/BrazilianSoccerMcpServer/Data/CsvLoader.cs
using System.Globalization;
using System.Text;
using BrazilianSoccerMcpServer.Models;
using CsvHelper;
using CsvHelper.Configuration;

namespace BrazilianSoccerMcpServer.Data;

public sealed class SoccerDataStore
{
    public List<Match> Matches { get; } = new();
    public List<Player> Players { get; } = new();

    public bool IsLoaded { get; set; }
}

public static class CsvLoader
{
    public static SoccerDataStore LoadFromDirectory(string directory)
    {
        var store = new SoccerDataStore();
        var dataPath = Path.Combine(directory, "data", "kaggle");

        LoadBrasileiraoMatches(store, Path.Combine(dataPath, "Brasileirao_Matches.csv"));
        LoadCopaDoBrasilMatches(store, Path.Combine(dataPath, "Brazilian_Cup_Matches.csv"));
        LoadLibertadoresMatches(store, Path.Combine(dataPath, "Libertadores_Matches.csv"));
        LoadBrFootballDataset(store, Path.Combine(dataPath, "BR-Football-Dataset.csv"));
        LoadNovoCampeonato(store, Path.Combine(dataPath, "novo_campeonato_brasileiro.csv"));
        LoadFifaPlayers(store, Path.Combine(dataPath, "fifa_data.csv"));

        store.IsLoaded = true;
        store.Matches.Sort((a, b) => Nullable.Compare(b.Date, a.Date));
        return store;
    }

    private static CsvConfiguration BaseConfig => new(CultureInfo.InvariantCulture)
    {
        HeaderValidated = null,
        MissingFieldFound = null,
        BadDataFound = null,
        TrimOptions = TrimOptions.Trim,
    };

    private static void LoadBrasileiraoMatches(SoccerDataStore store, string path)
    {
        if (!File.Exists(path)) return;

        var config = new CsvConfiguration(CultureInfo.InvariantCulture)
        {
            HeaderValidated = null,
            MissingFieldFound = null,
            BadDataFound = null,
            TrimOptions = TrimOptions.Trim,
        };

        using var reader = new StreamReader(path, Encoding.UTF8);
        using var csv = new CsvReader(reader, config);
        var records = csv.GetRecords<BrasileiraoMatchRecord>();
        foreach (var r in records)
        {
            if (!int.TryParse(r.season, out var season)) continue;
            if (!TryParseGoals(r.home_goal, out var homeGoals)) continue;
            if (!TryParseGoals(r.away_goal, out var awayGoals)) continue;

            var date = ParseDateTime(r.datetime);
            var home = r.home_team ?? string.Empty;
            var away = r.away_team ?? string.Empty;

            store.Matches.Add(new Match
            {
                MatchId = store.Matches.Count + 1,
                Date = date,
                HomeTeam = home,
                AwayTeam = away,
                NormalizedHomeTeam = NameNormalizer.Normalize(home),
                NormalizedAwayTeam = NameNormalizer.Normalize(away),
                HomeGoals = homeGoals,
                AwayGoals = awayGoals,
                Competition = "Brasileirão",
                Season = season,
                Round = r.round,
                HomeState = r.home_team_state,
                AwayState = r.away_team_state,
                Source = "Brasileirao_Matches.csv",
            });
        }
    }

    private static void LoadCopaDoBrasilMatches(SoccerDataStore store, string path)
    {
        if (!File.Exists(path)) return;

        using var reader = new StreamReader(path, Encoding.UTF8);
        using var csv = new CsvReader(reader, BaseConfig);
        var records = csv.GetRecords<CopaDoBrasilMatchRecord>();
        foreach (var r in records)
        {
            if (!int.TryParse(r.season, out var season)) continue;
            if (!TryParseGoals(r.home_goal, out var homeGoals)) continue;
            if (!TryParseGoals(r.away_goal, out var awayGoals)) continue;

            var date = ParseDateTime(r.datetime);
            var home = r.home_team ?? string.Empty;
            var away = r.away_team ?? string.Empty;

            store.Matches.Add(new Match
            {
                MatchId = store.Matches.Count + 1,
                Date = date,
                HomeTeam = home,
                AwayTeam = away,
                NormalizedHomeTeam = NameNormalizer.Normalize(home),
                NormalizedAwayTeam = NameNormalizer.Normalize(away),
                HomeGoals = homeGoals,
                AwayGoals = awayGoals,
                Competition = "Copa do Brasil",
                Season = season,
                Round = r.round,
                Source = "Brazilian_Cup_Matches.csv",
            });
        }
    }

    private static void LoadLibertadoresMatches(SoccerDataStore store, string path)
    {
        if (!File.Exists(path)) return;

        using var reader = new StreamReader(path, Encoding.UTF8);
        using var csv = new CsvReader(reader, BaseConfig);
        var records = csv.GetRecords<LibertadoresMatchRecord>();
        foreach (var r in records)
        {
            if (!int.TryParse(r.season, out var season)) continue;
            if (!TryParseGoals(r.home_goal, out var homeGoals)) continue;
            if (!TryParseGoals(r.away_goal, out var awayGoals)) continue;

            var date = ParseDateTime(r.datetime);
            var home = r.home_team ?? string.Empty;
            var away = r.away_team ?? string.Empty;

            store.Matches.Add(new Match
            {
                MatchId = store.Matches.Count + 1,
                Date = date,
                HomeTeam = home,
                AwayTeam = away,
                NormalizedHomeTeam = NameNormalizer.Normalize(home),
                NormalizedAwayTeam = NameNormalizer.Normalize(away),
                HomeGoals = homeGoals,
                AwayGoals = awayGoals,
                Competition = "Copa Libertadores",
                Season = season,
                Stage = r.stage,
                Source = "Libertadores_Matches.csv",
            });
        }
    }

    private static void LoadBrFootballDataset(SoccerDataStore store, string path)
    {
        if (!File.Exists(path)) return;

        using var reader = new StreamReader(path, Encoding.UTF8);
        using var csv = new CsvReader(reader, BaseConfig);
        var records = csv.GetRecords<BrFootballMatchRecord>();
        foreach (var r in records)
        {
            if (!TryParseGoals(r.home_goal, out var homeGoals)) continue;
            if (!TryParseGoals(r.away_goal, out var awayGoals)) continue;
            if (!int.TryParse(r.date?.Replace("-", "").Substring(0, 4), out var season) && !int.TryParse(r.date?.Split('-').FirstOrDefault(), out season))
            {
                if (!TryParseDate(r.date, out var parsedDate))
                    continue;
                season = parsedDate.Year;
            }

            var date = TryParseDate(r.date, out var d) ? d : (DateTime?)null;
            var home = r.home ?? string.Empty;
            var away = r.away ?? string.Empty;

            store.Matches.Add(new Match
            {
                MatchId = store.Matches.Count + 1,
                Date = date,
                HomeTeam = home,
                AwayTeam = away,
                NormalizedHomeTeam = NameNormalizer.Normalize(home),
                NormalizedAwayTeam = NameNormalizer.Normalize(away),
                HomeGoals = homeGoals,
                AwayGoals = awayGoals,
                Competition = r.tournament ?? "Brazilian Football",
                Season = season,
                HomeCorners = ParseDouble(r.home_corner),
                AwayCorners = ParseDouble(r.away_corner),
                HomeShots = ParseDouble(r.home_shots),
                AwayShots = ParseDouble(r.away_shots),
                HalfTimeResult = r.ht_result,
                Source = "BR-Football-Dataset.csv",
            });
        }
    }

    private static void LoadNovoCampeonato(SoccerDataStore store, string path)
    {
        if (!File.Exists(path)) return;

        using var reader = new StreamReader(path, Encoding.UTF8);
        using var csv = new CsvReader(reader, BaseConfig);
        var records = csv.GetRecords<NovoCampeonatoRecord>();
        foreach (var r in records)
        {
            if (!int.TryParse(r.Ano, out var season)) continue;
            if (!TryParseGoals(r.Gols_mandante, out var homeGoals)) continue;
            if (!TryParseGoals(r.Gols_visitante, out var awayGoals)) continue;

            var date = TryParseDate(r.Data, out var d) ? d : (DateTime?)null;
            var home = r.Equipe_mandante ?? string.Empty;
            var away = r.Equipe_visitante ?? string.Empty;

            store.Matches.Add(new Match
            {
                MatchId = store.Matches.Count + 1,
                Date = date,
                HomeTeam = home,
                AwayTeam = away,
                NormalizedHomeTeam = NameNormalizer.Normalize(home),
                NormalizedAwayTeam = NameNormalizer.Normalize(away),
                HomeGoals = homeGoals,
                AwayGoals = awayGoals,
                Competition = "Brasileirão",
                Season = season,
                Round = r.Rodada,
                HomeState = r.Mandante_UF,
                AwayState = r.Visitante_UF,
                Stadium = r.Arena,
                Source = "novo_campeonato_brasileiro.csv",
            });
        }
    }

    private static void LoadFifaPlayers(SoccerDataStore store, string path)
    {
        if (!File.Exists(path)) return;

        using var reader = new StreamReader(path, Encoding.UTF8);
        using var csv = new CsvReader(reader, BaseConfig);
        var records = csv.GetRecords<FifaPlayerRecord>();
        foreach (var r in records)
        {
            if (!int.TryParse(r.ID, out var id)) continue;

            store.Players.Add(new Player
            {
                Id = id,
                Name = r.Name ?? string.Empty,
                Age = int.TryParse(r.Age, out var age) ? age : 0,
                Nationality = r.Nationality ?? string.Empty,
                Overall = int.TryParse(r.Overall, out var overall) ? overall : 0,
                Potential = int.TryParse(r.Potential, out var potential) ? potential : 0,
                Club = r.Club ?? string.Empty,
                Position = r.Position ?? string.Empty,
                JerseyNumber = r.Jersey_Number,
                Height = r.Height,
                Weight = r.Weight,
                Crossing = NullableInt(r.Crossing),
                Finishing = NullableInt(r.Finishing),
                Dribbling = NullableInt(r.Dribbling),
                ShortPassing = NullableInt(r.ShortPassing),
            });
        }
    }

    private static bool TryParseGoals(string? value, out int? goals)
    {
        goals = null;
        if (string.IsNullOrWhiteSpace(value)) return false;
        if (int.TryParse(value, out var g))
        {
            goals = g;
            return true;
        }
        return false;
    }

    private static double? ParseDouble(string? value)
    {
        if (string.IsNullOrWhiteSpace(value)) return null;
        if (double.TryParse(value, NumberStyles.Any, CultureInfo.InvariantCulture, out var d))
            return d;
        return null;
    }

    private static int? NullableInt(string? value)
    {
        if (string.IsNullOrWhiteSpace(value)) return null;
        if (int.TryParse(value, out var i))
            return i;
        return null;
    }

    private static DateTime? ParseDateTime(string? value)
    {
        if (TryParseDate(value, out var date))
            return date;
        return null;
    }

    private static bool TryParseDate(string? value, out DateTime date)
    {
        date = default;
        if (string.IsNullOrWhiteSpace(value)) return false;

        var formats = new[]
        {
            "yyyy-MM-dd HH:mm:ss",
            "yyyy-MM-dd",
            "dd/MM/yyyy",
            "MM/dd/yyyy",
        };

        if (DateTime.TryParseExact(value.Trim(), formats, CultureInfo.InvariantCulture, DateTimeStyles.None, out date))
            return true;

        if (DateTime.TryParse(value, CultureInfo.InvariantCulture, DateTimeStyles.None, out date))
            return true;

        return false;
    }

    private sealed class BrasileiraoMatchRecord
    {
        public string? datetime { get; set; }
        public string? home_team { get; set; }
        public string? home_team_state { get; set; }
        public string? away_team { get; set; }
        public string? away_team_state { get; set; }
        public string? home_goal { get; set; }
        public string? away_goal { get; set; }
        public string? season { get; set; }
        public string? round { get; set; }
    }

    private sealed class CopaDoBrasilMatchRecord
    {
        public string? round { get; set; }
        public string? datetime { get; set; }
        public string? home_team { get; set; }
        public string? away_team { get; set; }
        public string? home_goal { get; set; }
        public string? away_goal { get; set; }
        public string? season { get; set; }
    }

    private sealed class LibertadoresMatchRecord
    {
        public string? datetime { get; set; }
        public string? home_team { get; set; }
        public string? away_team { get; set; }
        public string? home_goal { get; set; }
        public string? away_goal { get; set; }
        public string? season { get; set; }
        public string? stage { get; set; }
    }

    private sealed class BrFootballMatchRecord
    {
        public string? tournament { get; set; }
        public string? home { get; set; }
        public string? home_goal { get; set; }
        public string? away_goal { get; set; }
        public string? away { get; set; }
        public string? home_corner { get; set; }
        public string? away_corner { get; set; }
        public string? home_attack { get; set; }
        public string? away_attack { get; set; }
        public string? home_shots { get; set; }
        public string? away_shots { get; set; }
        public string? time { get; set; }
        public string? date { get; set; }
        public string? ht_diff { get; set; }
        public string? at_diff { get; set; }
        public string? ht_result { get; set; }
        public string? at_result { get; set; }
        public string? total_corners { get; set; }
    }

    private sealed class NovoCampeonatoRecord
    {
        public string? ID { get; set; }
        public string? Data { get; set; }
        public string? Ano { get; set; }
        public string? Rodada { get; set; }
        public string? Equipe_mandante { get; set; }
        public string? Equipe_visitante { get; set; }
        public string? Gols_mandante { get; set; }
        public string? Gols_visitante { get; set; }
        public string? Mandante_UF { get; set; }
        public string? Visitante_UF { get; set; }
        public string? Vencedor { get; set; }
        public string? Arena { get; set; }
        public string? OBS { get; set; }
    }

    private sealed class FifaPlayerRecord
    {
        public string? ID { get; set; }
        public string? Name { get; set; }
        public string? Age { get; set; }
        public string? Photo { get; set; }
        public string? Nationality { get; set; }
        public string? Flag { get; set; }
        public string? Overall { get; set; }
        public string? Potential { get; set; }
        public string? Club { get; set; }
        public string? Club_Logo { get; set; }
        public string? Value { get; set; }
        public string? Wage { get; set; }
        public string? Special { get; set; }
        public string? Preferred_Foot { get; set; }
        public string? International_Reputation { get; set; }
        public string? Weak_Foot { get; set; }
        public string? Skill_Moves { get; set; }
        public string? Work_Rate { get; set; }
        public string? Body_Type { get; set; }
        public string? Real_Face { get; set; }
        public string? Position { get; set; }
        public string? Jersey_Number { get; set; }
        public string? Joined { get; set; }
        public string? Loaned_From { get; set; }
        public string? Contract_Valid_Until { get; set; }
        public string? Height { get; set; }
        public string? Weight { get; set; }
        public string? Crossing { get; set; }
        public string? Finishing { get; set; }
        public string? HeadingAccuracy { get; set; }
        public string? ShortPassing { get; set; }
        public string? Volleys { get; set; }
        public string? Dribbling { get; set; }
    }
}
